File: solutions/csharp/proverb/1/Proverb.cs
using System;
using System.Collections.Generic;
using System.Linq;

public static class Proverb
{
	public static string[] Recite(string[] subjects)
	{
		if (!subjects.Any())
		{
			return Array.Empty<string>();
		}
		else
		{
			List<string> subjectList = subjects.ToList();
			string firstSubject = subjectList.Shift();

			List<string> verses = new();
			string previousSubject = firstSubject;
			while (subjectList.Any())
			{
				string currentSubject = subjectList.Shift();
				verses.Add($"For want of a {previousSubject} the {currentSubject} was lost.");
				previousSubject = currentSubject;
			}

			verses.Add($"And all for the want of a {firstSubject}.");
			return verses.ToArray();
		}
	}
}

public static class CollectionExtensions
{
	public static T Shift<T>(this ICollection<T> haystack)
	{
		T value = haystack.FirstOrDefault();
		_ = haystack.Remove(value);
		return value;
	}
}

File: solutions/python/proverb/1/proverb.py
def proverb(*pieces, qualifier=None):
    if not pieces:
        return []

    verses = []
    for index in range(len(pieces) - 1):
        verses.append(_build_consequence(pieces[index], pieces[index + 1]))

    verses.append(_build_final_verse(pieces[0], qualifier))
    return verses

def _build_consequence(lost_item, next_item):
    return f"For want of a {lost_item} the {next_item} was lost."

def _build_final_verse(first_item, qualifier):
    if qualifier:
        first_item = f"{qualifier} {first_item}"

    return f"And all for the want of a {first_item}."